src/AdvancedMH.jl

```diff
@@ -0,0 +1,135 @@
+using Reexport
+using Turing.Interface
+@reexport using Distributions
+using Random
+
+# Import specific functions and types to use or overload.
+import Distributions: VariateForm, ValueSupport, variate_form, value_support, Sampleable
+import MCMCChains: Chains
+import Turing.Interface: step!, AbstractSampler, AbstractTransition, transition_type
+
+# Exports
+export MetropolisHastings, DensityModel, sample
+
+# Define a sampler type.
+struct MetropolisHastings{T} <: AbstractSampler 
+    init_θ :: T
+    proposal :: Function
+end
+
+# Default constructors.
+MetropolisHastings(init_θ::T) where T<:Real = MetropolisHastings{T}(init_θ, x -> Normal(x, 1.0))
+MetropolisHastings(init_θ::T) where T<:Vector{<:Real} = MetropolisHastings{T}(init_θ, x -> MvNormal(x, 1.0))
+
+# Define a model type. Stores the log density function and the data to 
+# evaluate the log density on.
+struct DensityModel{V<:VariateForm, S<:ValueSupport, T} <: Sampleable{V, S}
+    π :: Function
+    data :: T
+end
+
+# Default density constructor.
+DensityModel(π::Function, data::T) where T = DensityModel{VariateForm, ValueSupport, T}(π, data)
+
+# Create a very basic Transition type, only stores the 
+# parameter draws and the log probability of the draw.
+struct Transition{T<:Union{Vector{<:Real}, <:Real}} <: AbstractTransition
+    θ :: T
+    lp :: Float64
+end
+
+# Store the new draw and its log density.
+Transition(model::M, θ::T) where {M<:DensityModel, T} = Transition(θ, ℓπ(model, θ))
+
+# Tell the interface what transition type we would like to use.
+transition_type(model::DensityModel, spl::MetropolisHastings) = typeof(Transition(spl.init_θ, ℓπ(model, spl.init_θ)))
+
+# Define a function that makes a basic proposal depending on a univariate
+# parameterization or a multivariate parameterization.
+proposal(spl::MetropolisHastings, model::DensityModel, θ::Real) = Transition(model, rand(spl.proposal(θ)))
+proposal(spl::MetropolisHastings, model::DensityModel, θ::Vector{<:Real}) = Transition(model, rand(spl.proposal(θ)))
+proposal(spl::MetropolisHastings, model::DensityModel, t::Transition) = proposal(spl, model, t.θ)
+
+# Calculate the logpdf of one proposal given another proposal.
+q(spl::MetropolisHastings, θ1::Real, θ2::Real) = logpdf(spl.proposal(θ1), θ2)
+q(spl::MetropolisHastings, θ1::Vector{<:Real}, θ2::Vector{<:Real}) = logpdf(spl.proposal(θ1), θ2)
+q(spl::MetropolisHastings, t1::Transition, t2::Transition) = q(spl, t1.θ, t2.θ)
+
+# Calculate the density of the model given some parameterization.
+ℓπ(model::DensityModel, θ::T) where T = model.π(model.data, θ)
+ℓπ(model::DensityModel, t::Transition) = t.lp
+
+# Define the first step! function, which is called at the 
+# beginning of sampling. Return the initial parameter used
+# to define the sampler.
+function step!(
+    rng::AbstractRNG,
+    model::M,
+    spl::S,
+    N::Integer;
+    kwargs...
+) where {
+    M <: DensityModel,
+    S <: MetropolisHastings
+}
+    return Transition(model, spl.init_θ)
+end
```

Comment:
This can probably be re-written to
```julia
function step!(
    rng::AbstractRNG,
    model::DensityModel,
    spl::MetropolisHastings,
    N::Integer,
    kwargs...,
)
    return Transition(model, spl.init_θ)
end
```
since the `where` doesn't add anything.